File: day4/State.py
class Interpreter:
    def __init__(self, src, tape):
        self.src = src
        self.tape = tape

    def checkBraceBalance(self) -> int:
        head = 0
        opens = 0
        closes = 0
        while(head < len(self.src)):
            if(self.src[head] == '['):
                opens += 1
            elif(self.src[head] == ']'):
                closes += 1

            if((opens - closes) < 0):
                return -1

            head += 1

        return opens - closes

    def interpret(self):
        balance = self.checkBraceBalance()
        if(balance > 0):
            print("[ERROR]: Unbalanced '['.")
            exit(1)
        elif(balance < 0):
            print("[ERROR]: Unbalanced ']'.")
            exit(1)

        head = 0
        while(head >= 0 and head < len(self.src)):
            match self.src[head]:
                case '>':
                    self.tape.headInc()
                case '<':
                    self.tape.headDec()
                case '+':
                    self.tape.incAtHead()
                case '-':
                    self.tape.decAtHead()
                case '.':
                    print(chr(self.tape.getVal()), end="")
                case ',':
                    self.tape.setVal(ord(input()[0]))
                case '[':
                    if(self.tape.getVal() == 0):
                        depth = 0
                        while(head < len(self.src)):
                            if(self.src[head] == '['):
                                depth += 1
                            elif(self.src[head] == ']'):
                                depth -= 1
                            
                            if(depth == 0):
                                break

                            head += 1
                        
                        if(head >= len(self.src)):
                            print("[ERROR]: Unbounded loop.")
                            exit(1)
                case ']':
                    if(self.tape.getVal() != 0):
                        depth = 0
                        while(head >= 0):
                            if(self.src[head] == '['):
                                depth += 1
                            elif(self.src[head] == ']'):
                                depth -= 1

                            if(depth == 0):
                                break

                            head -= 1
                        
                        if(head < 0):
                            print("[ERROR]: Unbounded loop.")
                            exit(1)
                case _:
                    pass
            
            head += 1

class Tape:
    def __init__(self):
        self.tape = []
        self.head = 0
        self.size = 256

        for i in range(self.size):
            self.tape.append(0)
    
    def headInc(self):
        self.head += 1

        if(self.head >= self.size):
            oldSize = self.size
            self.size *= 2
            for i in range(oldSize):
                self.tape.append(0)
    
    def headDec(self):
        if(self.head == 0):
            return
        
        self.head -= 1

    def getVal(self):
        return self.tape[self.head]

    def setVal(self, value):
        self.tape[self.head] = value

    def incAtHead(self):
        self.tape[self.head] = (self.tape[self.head] + 1) % 256

    def decAtHead(self):
        self.tape[self.head] -= 1
        if(self.tape[self.head] < 0):
            self.tape[self.head] = 255

File: day6/State.rb
class Tape
    def initialize
        @head = 0
        @size = 256
        @tape = Array.new(@size, 0)
    end

    def get_val
        return @tape[@head]
    end

    def head_inc
        @head += 1

        if @head >= @tape.length
            old_size = @size
            for i in 0..@size
                @tape.push(0)
            end
            @size *= 2
        end
    end

    def head_dec
        if @head < 0
            return
        end

        @head -= 1
    end

    def inc_at_head
        @tape[@head] = (@tape[@head] + 1) % 256
    end

    def dec_at_head
        temp = @tape[@head] - 1
        if temp < 0
            @tape[@head] = 255
        else
            @tape[@head] -= 1
        end
    end

    def emit
        print(@tape[@head].chr)
    end

    def input
        @tape[@head] = gets.chars[0].ord
    end
end

class Interpreter
    def initialize
        @tape = Tape.new
    end

    def repl
        # repl
        puts "Brainfuck Interpreter\n\nType 'bye' to exit."
        while true
            print("$ ")
            STDOUT.flush

            line = STDIN.gets.chomp
            if line == "bye"
                exit(0)
            end

            interpret(line)
        end
    end

    def _check_balance(src)
        opens = 0
        closes = 0

        src_list = src.chars
        src_list.each do |head|
            case head
            when '['
                opens += 1
            when ']'
                closes += 1
            end

            if closes > opens
                break
            end
        end

        balance = opens - closes
        if balance > 0
            puts "[ERROR]: Unbalanced '['."
            exit(1)
        elsif balance < 0
            puts "[ERROR]: Unbalanced ']'."
            exit(1)
        end
    end

    def interpret(src)
        _check_balance(src)

        src_list = src.chars
        head = 0
        while head < src_list.length
            case src_list[head]
            when '>'
                @tape.head_inc
            when '<'
                @tape.head_dec
            when '+'
                @tape.inc_at_head
            when '-'
                @tape.dec_at_head
            when '.'
                @tape.emit
            when ','
                @tape.input
            when '['
                if @tape.get_val == 0
                    depth = 0
                    while head < src_list.length
                        if src_list[head] == '['
                            depth += 1
                        elsif src_list[head] == ']'
                            depth -= 1
                        end

                        if depth == 0
                            break
                        end

                        head += 1
                    end
                end
            when ']'
                if @tape.get_val != 0
                    depth = 0
                    while head > 0
                        if src_list[head] == '['
                            depth += 1
                        elsif src_list[head] == ']'
                            depth -= 1
                        end

                        if depth == 0
                            break
                        end

                        head -= 1
                    end
                end
            end

            head += 1
        end
    end
end